pdata/internal/metrics.go

```diff
@@ -185,42 +185,69 @@ func (at MetricAggregationTemporality) String() string {
 
 // MetricDataPointFlags defines how a metric aggregator reports aggregated values.
 // It describes how those values relate to the time interval over which they are aggregated.
-type MetricDataPointFlags uint32
+//
+// This is a reference type, if passed by value and callee modifies it the
+// caller will see the modification.
+//
+// Must use NewMetricDataPointFlags function to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type MetricDataPointFlags struct {
+	orig *uint32
+}
 
-const (
-	// MetricDataPointFlagsNone is the default MetricDataPointFlags.
-	MetricDataPointFlagsNone = MetricDataPointFlags(otlpmetrics.DataPointFlags_FLAG_NONE)
-)
+func newMetricDataPointFlags(orig *uint32) MetricDataPointFlags {
+	return MetricDataPointFlags{orig: orig}
+}
 
-// NewMetricDataPointFlags returns a new MetricDataPointFlags combining the flags passed
-// in as parameters.
-func NewMetricDataPointFlags(flags ...MetricDataPointFlag) MetricDataPointFlags {
-	var flag MetricDataPointFlags
-	for _, f := range flags {
-		flag |= MetricDataPointFlags(f)
-	}
-	return flag
+// NewMetricDataPointFlags creates a new empty NewMetricDataPointFlags.
+//
+// This must be used only in testing code. Users should use "AppendEmpty" when part of a Slice,
+// OR directly access the member if this is embedded in another struct.
+func NewMetricDataPointFlags() MetricDataPointFlags {
+	return newMetricDataPointFlags(new(uint32))
 }
 
-// HasFlag returns true if the MetricDataPointFlags contains the specified flag
-func (d MetricDataPointFlags) HasFlag(flag MetricDataPointFlag) bool {
-	return d&MetricDataPointFlags(flag) != 0
+// MoveTo moves all properties from the current struct to dest
+// resetting the current instance to its zero value
+func (ms MetricDataPointFlags) MoveTo(dest MetricDataPointFlags) {
+	*dest.orig = *ms.orig
+	ms.Reset()
 }
 
-// String returns the string representation of the MetricDataPointFlags.
-func (d MetricDataPointFlags) String() string {
-	return otlpmetrics.DataPointFlags(d).String()
+// CopyTo copies all properties from the current struct to the dest.
+func (ms MetricDataPointFlags) CopyTo(dest MetricDataPointFlags) {
+	*dest.orig = *ms.orig
+}
+
+// None returns true if MetricDataPointFlags is FLAG_NONE.
+func (ms MetricDataPointFlags) None() bool {
+	return *ms.orig == uint32(otlpmetrics.DataPointFlags_FLAG_NONE)
 }
 
-// MetricDataPointFlag allow users to configure DataPointFlags. This is achieved via NewMetricDataPointFlags.
-// The separation between MetricDataPointFlags and MetricDataPointFlag exists to prevent users accidentally
-// comparing the value of individual flags with MetricDataPointFlags. Instead, users must use the HasFlag method.
-type MetricDataPointFlag uint32
+// Reset resets the flags to be FLAG_NONE.
+func (ms MetricDataPointFlags) Reset() {
+	*ms.orig = uint32(otlpmetrics.DataPointFlags_FLAG_NONE)
+}
```

Comment:
My thought for adding them was to try to keep some backwards compatible capabilities and to future-proof the API.

The API today exposes `MetricDataPointFlagsNone` which can be used to compared against `Flags()` to see if no flags are set.  Without a `None()` function we'll only be able to check whether or not `NoRecordedValue` is set or not.  I couldn't find any use of that exposed field in Contrib so maybe it doesn't matter but it is technically possible today.

Since that is the only flag right now a return of false from `NoRecordedValue` means that no flags are set, but that wouldn't necessarily be true if more flags are added in the future. It feels like being able to interact with the default state of the enum via dedicated functions future proofs the API.

Reply:
Although I do think `None()` is good ergonomics for checking if there are no flags set, I do recognize that is not being done anywhere in Core or Contrib.  If a third flag is ever added to the protos and pdata is updated to handle that flag then we should add `None()` and `Rest()`.